### DISCUSSION TITLE
Help With BlackPill STM32 F411

### DISCUSSION THREAD

**Discussion by NRKorsakov** [shanteacontrols/OpenDeck#285] [General]
Hi! I bought a few of the BlackPills to be the brain of a midi controller. I'm having problems flashing the board.

I tried OpenDeck previosuly on a Raspberry PI Pico just to learn, and it worked just fine, but I wanted more ins & outs.

The Blackpill uses a different method to flash, and I can't get it to work?

I plug the BlackPill into my MacBook Pro 2013 running Big Sur.

To get into Bootloader-Mode I follow the instructions;

1	Connect the board to computer with USB C cable
	2	Hold down BOOT0 button
	3	While holding down BOOT0 button, press and hold NRST button for 1 second
	4	Release NRST button while still holding BOOT0 button for another 1 second
	5	Release BOOT0 button

The Blackpill looks like this;

![IMG_1025](https://github.com/shanteacontrols/OpenDeck/assets/25620049/0cf51a69-8e8f-43d1-9e07-b9cadb5b2b79)

I then open up Terminal and paste in:

sudo dfu-util -l

Terminal gives me an error message:

<img width="591" alt="Screenshot 2024-04-04 at 09 39 37 AM" src="https://github.com/shanteacontrols/OpenDeck/assets/25620049/aec5a3d8-1891-45ea-a059-4419f1cf97d0">

Am I missing a step, or is there another method of flashing the Blackpill?

Thanks!

**Comment by NRKorsakov** [comment]
I think I remember using the Arduino software to upload OpenDeck to the Pico?

Can I use it for the BlackPill?

That was much easier for us tech-challenged people.

Actually just looked and it looks like it was just downloading the latest OpenDeck release for the Raspberry Pi Pico and dragging the file onto the board in OSX's finder, if I remember correctly?

**Comment by paradajz** [comment]
The wiki clearly states the supported operating systems for the flashing of STM32.
https://github.com/shanteacontrols/OpenDeck/wiki/Flashing-STM32-boards

macOS isn't among them.

**Reply by basst22778** [reply]
But it used to be - there was a documentation for flashing with macOS and homebrew in the wiki
Flashed a F411 Blackpill yesterday with the latest firmware on my intel Mac.

**Reply by NRKorsakov** [reply]
Hi thanks for that!

Can I ask how you did it?

I started poking around and found this;

https://www.sgbotic.com/index.php?dispatch=pages.view&page_id=49

**Reply by basst22778** [reply]
There used to be a manual in the opendeck repository. 
Maybe you can find it by checking out an old version (6.x) locally?

**Reply by NRKorsakov** [reply]
Hi thank you for that! I will look for it that way.

I tried the method at the link above, but so far no luck, but admittedly I only had about 1/2 hour to try it, and I am totally non-tech-savvy.

I did various things to try to see how to flash a BlackPill with a .bin or .hex file. I tried the above link method with one or the other, I don't remember but I think it might have been .hex?

Th Arduino IDE says it was "including" it in the upload to the board, or whatever that language is, and it said "successful", but when I unplug and then plug back in the BlackPill it does not show up as a midi device.

I'll look for the instructions you mention, and if worse comes to worse, I can just put two or three Pico's inside the midi controller along with a little USB hub, and that would get me to the number of pots I need.

If I knew anything about this stuff I know I could use a "mix" or whatever, but that is way beyond me.

I did love it when the Pico worked, as it is solution accessible for us musicians who didn't grow up in the age of coding, it works so easily like the Korg Nano's I am replacing, just plug it in, a simple web application tells it what to do and saves it to memory, and then back to making music!

I would love to learn something about coding some day, but that's just not in the cards.

Anyway thanks for your help!

**Reply by NRKorsakov** [reply]
That is "mux", not "mix", thank you spellcheck!

**Comment by NRKorsakov** [comment]
It its confusing because at this link there are instructions for flashing STM32 boards from a Mac;

https://github.com/shanteacontrols/OpenDeck/wiki/Flashing-STM32-boards/a5b44b07880030af99308d049b301aa182168c49#macos

And it says;

___________________________

macOS

Note: M1 macs aren't tested.

If the instructions [from this page](https://github.com/shanteacontrols/OpenDeck/wiki/Building-the-OpenDeck-firmware#getting-the-packages) have already been followed and executed, this section should be skipped.

It is assumed here macOS Catalina or higher is used. Earlier versions are not supported. Run the following commands from terminal to download the necessary packages (assuming [Homebrew](https://brew.sh/) is installed).

Run the following commands from terminal to download the necessary packages (select the entire block and paste in into terminal):

brew install dfu-util
___________________________

And no my MacBook is not an M1, it's about 11 years old I think.

**Reply by paradajz** [reply]
I can no longer offer support for macOS because I don't have any Apple device. I also do not use Windows. Therefore, the development environment is now supported only on Linux. You can of course make it work on both macOS and Windows with some effort but I do not offer advice/help/support on that because I don't use it. I think the current flashing guide is really clear and simple.

**Comment by basst22778** [comment]
I will try to install ubuntu in a Virtual Machine on macOS the next days because i didn‘t succeed with the build pipeline on macOS (with my knowledge) 

if i am successful i can share the vm image for other macOS users.

https://ubuntu.com/blog/how-to-create-a-vscode-linux-remote-environment

**Reply by NRKorsakov** [reply]
Hi thanks that would be really helpful!

And much appreciated!

I'm glad there is a discussion group for this project.

Thanks again!

**Reply by basst22778** [reply]
you can install ubuntu as a vm in Visual Studio Code. But this is only useful for building custom firmware.
If only you want to flash STM32 on MacOS, i succeeded with homebrew.

https://brew.sh

> /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

after installing homebrew, you can install the dfu-util port for mac.

> brew install dfu-util

than you have to follow the wiki: 
> sudo dfu-util -l

in an old version of the wiki i found the command to flash the binaries from the repository without make 

The binary is included in the actual release on github
https://github.com/shanteacontrols/OpenDeck/releases

> dfu-util -d 0483:df11 -a 0 -i 0 -s 0x8000000:leave -D blackpill411.bin

**Comment by NRKorsakov** [comment]
I bought something called an "ST-Link" ($7.00 on Amazon) that is supposedly, from what I read, the way that the Mac can program the BlackPill.

I'll post something if it works.

**Comment by NRKorsakov** [comment]
Hey for all who are interested, the little ST LINK V2 USB "dongle" sort of thing worked!

You plug the four (programming pins? I don't know what they are called?) on the end of the BlackPill into the ST LINK pins (there are guides if you Google it), then plug the ST LINK into your computer (I used a MacBook).

Then you download something called the STM32CubeProgrammer from the STM32 site. There is something called the STM32TrustedPackageCreator that downloads too, but I used the Programmer only.

Then you download the OpenDeck BlackPill 411 .bin file from this GitHub (I didn't try the .hex file, but the .bin worked for me), in the Programmer app, you hit the "connect" button to the ST LINK, In the Erasing & Programming Tab in the Programmer choose the OpenDeck BlackPill .bin file you downloaded, and hit the programming button. There are two programming buttons, one regular one and one marked Automatic. The Automatic didn't work for me. It will ("flash"?) OpenDeck to the BlackPill.

Then you can disconnect the ST LINK, plug the BlackPill in with a normal USB Cable, and it will show up in both AudioMidi Setup  and in the OpenDeck Configurator. I used the latter in Chrome, not Safari, as Safari can sometimes be weird with stuff like that.

I haven't had time to try it yet with actual buttons and pots connected, but it looks good so far!

I'll update if there are any problems.

**Reply by ianchia** [reply]
I can confirm that @NRKorsakov's technique worked perfectly with Apple Silicon using a Mac Studio (M2 Max), with the STM32CubeProgrammer running under macOS Ventura 13.6.8 and a generic ST-Link V2 from Amazon (I bought this one from Amazon AU https://www.amazon.com.au/dp/B0BQ6X4CYF). 

I used the [v7.10.0](https://github.com/shanteacontrols/OpenDeck/releases/tag/v7.10.0) build of the ([blackpill411.bin](https://github.com/shanteacontrols/OpenDeck/releases/download/v7.10.0/blackpill411.bin) with a [WeAct Black Pill V3.1 Board off AliExpress](https://www.aliexpress.com/item/1005001456186625.html). My Black Pill board is a 25MHz crystal HSE 8M Flash (WeAct says "Regular version, the external crystal is 25M and the back of the board is welded with w25q64jvssiq 8m flash". Note that the official [OpenDeck Readme page](https://github.com/shanteacontrols/OpenDeck?tab=readme-ov-file) says "WeAct (developer of Black Pill boards) currently sells Black Pill board with AT32F403ACGU7 MCU instead of STM32F411. This variant isn't supported!" but the board I bought from WeAct off AliExpress has a legit STM32F411CEU6 on it which STM32CubeProgrammer confirms. It's the WeAct Black Pill V3.1.Board.

Audio MIDI Setup shows the board appearing as "OpenDeck | blackpill411" and under Chrome on macOS Ventura,  OpenDeck Configurator shows "Board BlackPill F411 Firmwarev 7.10.0 Preset 1" and all the presets appear to work. I need to hook up some hardware to the board to confirm but I'm confident everything will be fine.

**Reply by NRKorsakov** [reply]
That is great news! And WOW, nobody has EVER cited me as someone who could help anyone with anything to do with computers or electronics! I can barely make my coffee-maker work? Anyway I'm glad this worked out. I flashed about 4 of those boards I got on AliExpress and have slowly been integrating them into my studios. They work great, but I'm still getting the hang of soldering with these shaky hands. Once again great thanks to and for Shantea Controls and Opendeck for this absolutely terrific (and very generous) resource for musicians.
